src/adapter/src/coord/sequencer/inner.rs

```
@@ -3577,32 +3555,200 @@ impl Coordinator {
+
+                        let catalog = coord.catalog().for_session(session);
+
+                        // Resolve items in statement
+                        mz_sql::names::resolve(&catalog, create_source_stmt)
+                            .map_err(|e| AdapterError::internal(ALTER_SOURCE, e))
+                    };
+
+                let (mut create_source_stmt, mut depends_on) =
+                    create_sql_to_stmt_deps(self, session, cur_entry.create_sql())?;
+
+                // Ensure that we are only dropping items on which we depend.
+                for t in &to_drop {
+                    // Remove dependency.
+                    let existed = depends_on.remove(t);
+                    if !existed {
+                        Err(AdapterError::internal(
+                            ALTER_SOURCE,
+                            format!("removed {t}, but {id} did not have dependency"),
+                        ))?;
+                    }
+                }
+
+                // We are doing a lot of unwrapping, so just make an error to reference; all of
+                // these invariants are guaranteed to be true because of how we plan subsources.
+                let purification_err =
+                    || AdapterError::internal(ALTER_SOURCE, "error in subsource purification");
+
+                let referenced_subsources = match create_source_stmt
+                    .referenced_subsources
+                    .as_mut()
+                    .ok_or(purification_err())?
+                {
+                    ReferencedSubsources::SubsetTables(ref mut s) => s,
+                    _ => return Err(purification_err()),
+                };
+
+                // Fixup referenced_subsources
+                referenced_subsources.retain(|CreateSourceSubsource { subsource, .. }| {
+                    match subsource
+                        .as_ref()
+                        .unwrap_or_else(|| panic!("{}", purification_err().to_string()))
```

Can you add a comment indicating how this is unreachable/why we panic instead of returning an error?

---

Done--it's just that retain is infallible and erroring rather than panicking is a nice to have. If there are any errors in this, we have a serious programming error.